File: WidgetSDK/WidgetInfo/src/main/java/com/proxy/service/widget/info/view/CsRoundLayout.java
package com.proxy.service.widget.info.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Path;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.widget.RelativeLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.proxy.service.widget.info.R;

/**
 * @author: cangHX
 * @date: 2023/11/20 11:47
 * @desc:
 */
public class CsRoundLayout extends RelativeLayout {

    private final RectF rectF = new RectF();
    private final float[] radiusF = new float[8];
    private final Path path = new Path();

    private float radiusLT = 0f;
    private float radiusRT = 0f;
    private float radiusLB = 0f;
    private float radiusRB = 0f;

    private int maxHeight = -1;

    public CsRoundLayout(@NonNull Context context) {
        this(context, null);
    }

    public CsRoundLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CsRoundLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    private void init(@NonNull Context context, @Nullable AttributeSet attrs) {
        if (isInLayout()) {
            return;
        }
        if (isInEditMode()) {
            return;
        }
        if (attrs == null) {
            return;
        }
        @SuppressLint("CustomViewStyleable") TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.cs_widget_RoundLayout, 0, 0);
        int radius = array.getDimensionPixelSize(R.styleable.cs_widget_RoundLayout_round_radius, -1);
        int radius_l_t = array.getDimensionPixelSize(R.styleable.cs_widget_RoundLayout_round_radius_top_left, -1);
        int radius_r_t = array.getDimensionPixelSize(R.styleable.cs_widget_RoundLayout_round_radius_top_right, -1);
        int radius_r_b = array.getDimensionPixelSize(R.styleable.cs_widget_RoundLayout_round_radius_bottom_right, -1);
        int radius_l_b = array.getDimensionPixelSize(R.styleable.cs_widget_RoundLayout_round_radius_bottom_left, -1);
        array.recycle();

        if (radius >= 0) {
            radiusLT = radius;
            radiusRT = radius;
            radiusLB = radius;
            radiusRB = radius;
        }

        if (radius_l_t >= 0) {
            radiusLT = radius_l_t;
        }

        if (radius_r_t >= 0) {
            radiusRT = radius_r_t;
        }

        if (radius_r_b >= 0) {
            radiusRB = radius_r_b;
        }

        if (radius_l_b >= 0) {
            radiusLB = radius_l_b;
        }

        radiusF[0] = radiusLT;
        radiusF[1] = radiusLT;
        radiusF[2] = radiusRT;
        radiusF[3] = radiusRT;
        radiusF[4] = radiusRB;
        radiusF[5] = radiusRB;
        radiusF[6] = radiusLB;
        radiusF[7] = radiusLB;
    }

    public void setRadius(int radius) {
        setRadius(radius, radius, radius, radius);
    }

    public void setRadius(int radiusLT, int radiusRT, int radiusRB, int radiusLB) {
        radiusF[0] = radiusLT;
        radiusF[1] = radiusLT;
        radiusF[2] = radiusRT;
        radiusF[3] = radiusRT;
        radiusF[4] = radiusRB;
        radiusF[5] = radiusRB;
        radiusF[6] = radiusLB;
        radiusF[7] = radiusLB;
        requestLayout();
        postInvalidate();
    }

    public void setMaxHeight(int maxHeight) {
        this.maxHeight = maxHeight;
        requestLayout();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int heightSpec = heightMeasureSpec;
        if (maxHeight > 0) {
            heightSpec = MeasureSpec.makeMeasureSpec(maxHeight, MeasureSpec.AT_MOST);
        }
        super.onMeasure(widthMeasureSpec, heightSpec);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        rectF.left = 0f;
        rectF.top = 0f;
        rectF.right = r - l;
        rectF.bottom = b - t;

        path.reset();
        path.addRoundRect(rectF, radiusF, Path.Direction.CCW);
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        int num = canvas.save();
        canvas.clipPath(path);
        super.dispatchDraw(canvas);
        canvas.restoreToCount(num);
    }

    @Override
    public void onDrawForeground(@NonNull Canvas canvas) {
        int num = canvas.save();
        canvas.clipPath(path);
        super.onDrawForeground(canvas);
        canvas.restoreToCount(num);
    }

    @Override
    public void draw(Canvas canvas) {
        int num = canvas.save();
        canvas.clipPath(path);
        super.draw(canvas);
        canvas.restoreToCount(num);
    }

//    @Override
//    protected void onDraw(Canvas canvas) {
//        int num = canvas.save();
//        canvas.clipPath(path);
//        super.onDraw(canvas);
//        canvas.restoreToCount(num);
//    }
}

File: WidgetSDK/WidgetInfo/src/main/java/com/proxy/service/widget/info/view/CsRoundLayout.kt
package com.proxy.service.widget.info.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Outline
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import android.view.ViewOutlineProvider
import com.proxy.service.core.framework.app.resource.CsDpUtils
import com.proxy.service.widget.info.R

/**
 * @author: cangHX
 * @date: 2023/11/20 11:47
 * @desc: 圆角 layout
 */
class CsRoundLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
    defStyleRes: Int = 0
) : CsMaxHeightLayout(context, attrs, defStyleAttr, defStyleRes) {

    private val rectF = RectF()
    private val radiusF = FloatArray(8)
    private val path = Path()

    private var pathDirty = true
    private var isUniformRadius = true
    private var uniformRadius = 0f

    private val maskPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_IN)
    }

    private val roundOutlineProvider = object : ViewOutlineProvider() {
        override fun getOutline(view: View, outline: Outline) {
            outline.setRoundRect(0, 0, view.width, view.height, uniformRadius)
        }
    }

    init {
        setWillNotDraw(false)
        init(context, attrs)
    }

    private fun init(context: Context, attrs: AttributeSet?) {
        if (attrs == null) {
            return
        }
        val array = context.obtainStyledAttributes(attrs, R.styleable.CsRoundLayout)

        val radius = getDimenPixelSize(array, R.styleable.CsRoundLayout_round_radius)
        val rTl = getDimenPixelSize(array, R.styleable.CsRoundLayout_round_radius_top_left)
        val rTr = getDimenPixelSize(array, R.styleable.CsRoundLayout_round_radius_top_right)
        val rBr = getDimenPixelSize(array, R.styleable.CsRoundLayout_round_radius_bottom_right)
        val rBl = getDimenPixelSize(array, R.styleable.CsRoundLayout_round_radius_bottom_left)
        array.recycle()

        var finalTL = 0f
        var finalTR = 0f
        var finalBR = 0f
        var finalBL = 0f

        if (radius >= 0) {
            finalTL = radius.toFloat()
            finalTR = radius.toFloat()
            finalBR = radius.toFloat()
            finalBL = radius.toFloat()
        }

        if (rTl >= 0) {
            finalTL = rTl.toFloat()
        }

        if (rTr >= 0) {
            finalTR = rTr.toFloat()
        }

        if (rBr >= 0) {
            finalBR = rBr.toFloat()
        }

        if (rBl >= 0) {
            finalBL = rBl.toFloat()
        }

        radiusF[0] = finalTL
        radiusF[1] = finalTL
        radiusF[2] = finalTR
        radiusF[3] = finalTR
        radiusF[4] = finalBR
        radiusF[5] = finalBR
        radiusF[6] = finalBL
        radiusF[7] = finalBL

        isUniformRadius = isUniform()
        uniformRadius = if (isUniformRadius) radiusF[0] else 0f
    }

    /**
     * 设置圆角
     * */
    fun setRadiusDp(radius: Float) {
        val r = CsDpUtils.dp2pxf(radius)
        updateRadiusF(r, r, r, r)
    }

    /**
     * 设置圆角
     * */
    fun setRadiusPx(radius: Int) {
        setRadiusPx(radius, radius, radius, radius)
    }

    /**
     * 设置圆角
     * */
    fun setRadiusDp(rTl: Float, rTr: Float, rBr: Float, rBl: Float) {
        updateRadiusF(
            CsDpUtils.dp2pxf(rTl),
            CsDpUtils.dp2pxf(rTr),
            CsDpUtils.dp2pxf(rBr),
            CsDpUtils.dp2pxf(rBl)
        )
    }

    /**
     * 设置圆角
     * */
    fun setRadiusPx(rTl: Int, rTr: Int, rBr: Int, rBl: Int) {
        updateRadiusF(rTl.toFloat(), rTr.toFloat(), rBr.toFloat(), rBl.toFloat())
    }


    private fun updateRadiusF(rTl: Float, rTr: Float, rBr: Float, rBl: Float) {
        radiusF[0] = rTl
        radiusF[1] = rTl
        radiusF[2] = rTr
        radiusF[3] = rTr
        radiusF[4] = rBr
        radiusF[5] = rBr
        radiusF[6] = rBl
        radiusF[7] = rBl
        pathDirty = true
        rebuildPathIfNeeded()
        postInvalidate()
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        if (changed) {
            pathDirty = true
        }
        rebuildPathIfNeeded()
    }

    private fun rebuildPathIfNeeded() {
        if (!pathDirty || width == 0 || height == 0) {
            return
        }
        pathDirty = false
        rectF.set(0f, 0f, width.toFloat(), height.toFloat())
        path.reset()
        path.addRoundRect(rectF, radiusF, Path.Direction.CCW)
        applyClippingMode()
    }

    private fun isUniform(): Boolean {
        val r = radiusF[0]
        return radiusF[2] == r && radiusF[4] == r && radiusF[6] == r
    }

    private fun applyClippingMode() {
        isUniformRadius = isUniform()
        uniformRadius = if (isUniformRadius) radiusF[0] else 0f

        if (isUniformRadius && uniformRadius > 0f) {
            outlineProvider = roundOutlineProvider
            clipToOutline = true
        } else {
            outlineProvider = ViewOutlineProvider.BACKGROUND
            clipToOutline = false
        }
        invalidateOutline()
    }

    override fun draw(canvas: Canvas) {
        if (isUniformRadius) {
            super.draw(canvas)
            return
        }
        val count = canvas.saveLayer(rectF, null)
        super.draw(canvas)
        canvas.drawPath(path, maskPaint)
        canvas.restoreToCount(count)
    }
}
